### PR TITLE
[bugfix][distributed] fix multi-node bug for shared memory

### LINKED ISSUE #6564: [Bug]: Distributed Inference and Serving
### Your current environment

```text
The output of `python collect_env.py`
```
Collecting environment information...
/model/vllm/vllm/usage/usage_lib.py:19: RuntimeWarning: Failed to read commit hash:
No module named 'vllm.commit_id'
  from vllm.version import __version__ as VLLM_VERSION
PyTorch version: 2.3.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.30.0
Libc version: glibc-2.35

Python version: 3.11.9 (main, Apr 19 2024, 16:48:06) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-116-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB
GPU 4: Tesla V100-SXM2-32GB
GPU 5: Tesla V100-SXM2-32GB
GPU 6: Tesla V100-SXM2-32GB
GPU 7: Tesla V100-SXM2-32GB

Nvidia driver version: 535.183.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.2.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.2.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               96
On-line CPU(s) list:                  0-95
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Gold 6271C CPU @ 2.60GHz
CPU family:                           6
Model:                                85
Thread(s) per core:                   2
Core(s) per socket:                   24
Socket(s):                            2
Stepping:                             7
BogoMIPS:                             5200.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp_epp pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            1.5 MiB (48 instances)
L1i cache:                            1.5 MiB (48 instances)
L2 cache:                             48 MiB (48 instances)
L3 cache:                             66 MiB (2 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-23,48-71
NUMA node1 CPU(s):                    24-47,72-95
Vulnerability Gather data sampling:   Mitigation; Microcode
Vulnerability Itlb multihit:          KVM: Mitigation: VMX disabled
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-nccl-cu12==2.20.5
[pip3] torch==2.3.0
[pip3] torchvision==0.18.0
[pip3] transformers==4.42.4
[pip3] triton==2.3.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] nvidia-nccl-cu12          2.20.5                   pypi_0    pypi
[conda] torch                     2.3.0                    pypi_0    pypi
[conda] torchvision               0.18.0                   pypi_0    pypi
[conda] transformers              4.42.4                   pypi_0    pypi
[conda] triton                    2.3.0                    pypi_0    pypi
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.5.2
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
GPU0    GPU1    GPU2    GPU3    GPU4    GPU5    GPU6    GPU7    NIC0    NIC1    NIC2    NIC3    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      NV1     NV2     NV1     SYS     SYS     SYS     NV2     NODE    NODE    SYS     SYS     0-23,48-71      0               N/A
GPU1    NV1      X      NV1     NV2     SYS     SYS     NV2     SYS     NODE    NODE    SYS     SYS     0-23,48-71      0               N/A
GPU2    NV2     NV1      X      NV2     SYS     NV1     SYS     SYS     PIX     PIX     SYS     SYS     0-23,48-71      0               N/A
GPU3    NV1     NV2     NV2      X      NV1     SYS     SYS     SYS     PIX     PIX     SYS     SYS     0-23,48-71      0               N/A
GPU4    SYS     SYS     SYS     NV1      X      NV2     NV2     NV1     SYS     SYS     PIX     PIX     24-47,72-95     1               N/A
GPU5    SYS     SYS     NV1     SYS     NV2      X      NV1     NV2     SYS     SYS     PIX     PIX     24-47,72-95     1               N/A
GPU6    SYS     NV2     SYS     SYS     NV2     NV1      X      NV1     SYS     SYS     NODE    NODE    24-47,72-95     1               N/A
GPU7    NV2     SYS     SYS     SYS     NV1     NV2     NV1      X      SYS     SYS     NODE    NODE    24-47,72-95     1               N/A
NIC0    NODE    NODE    PIX     PIX     SYS     SYS     SYS     SYS      X      PIX     SYS     SYS
NIC1    NODE    NODE    PIX     PIX     SYS     SYS     SYS     SYS     PIX      X      SYS     SYS
NIC2    SYS     SYS     SYS     SYS     PIX     PIX     NODE    NODE    SYS     SYS      X      PIX
NIC3    SYS     SYS     SYS     SYS     PIX     PIX     NODE    NODE    SYS     SYS     PIX      X 

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_0
  NIC1: mlx5_1
  NIC2: mlx5_2
 NIC3: mlx5_3

### 🐛 Describe the bug

vllm) root@ubuntu:/model# tail -f nohup_output.log 
INFO 07-19 06:57:49 api_server.py:212] vLLM API server version 0.5.2
INFO 07-19 06:57:49 api_server.py:213] args: Namespace(host=None, port=xxxxxxx, uvicorn_log_level='info', allow_credentials=False, allowed_origins=['*'], allowed_methods=['*'], allowed_headers=['*'],, lora_modules=None, prompt_adapters=None, chat_template=None, response_role='assistant', ssl_keyfile=None, ssl_certfile=None, ssl_ca_certs=None, ssl_cert_reqs=0, root_path=None, middleware=[], model='/model/models/Qwen2-72B-Instruct', tokenizer=None, skip_tokenizer_init=False, revision=None, code_revision=None, tokenizer_revision=None, tokenizer_mode='auto', trust_remote_code=True, download_dir=None, load_format='auto', dtype='float16', kv_cache_dtype='auto', quantization_param_path=None, max_model_len=None, guided_decoding_backend='outlines', distributed_executor_backend='ray', worker_use_ray=False, pipeline_parallel_size=1, tensor_parallel_size=16, max_parallel_loading_workers=None, ray_workers_use_nsight=False, block_size=16, enable_prefix_caching=False, disable_sliding_window=False, use_v2_block_manager=False, num_lookahead_slots=0, seed=0, swap_space=8, gpu_memory_utilization=0.95, num_gpu_blocks_override=None, max_num_batched_tokens=None, max_num_seqs=256, max_logprobs=20, disable_log_stats=False, quantization=None, rope_scaling=None, rope_theta=None, enforce_eager=True, max_context_len_to_capture=None, max_seq_len_to_capture=8192, disable_custom_all_reduce=True, tokenizer_pool_size=0, tokenizer_pool_type='ray', tokenizer_pool_extra_config=None, enable_lora=False, max_loras=1, max_lora_rank=16, lora_extra_vocab_size=256, lora_dtype='auto', long_lora_scaling_factors=None, max_cpu_loras=None, fully_sharded_loras=False, enable_prompt_adapter=False, max_prompt_adapters=1, max_prompt_adapter_token=0, device='auto', scheduler_delay_factor=0.0, enable_chunked_prefill=False, speculative_model=None, num_speculative_tokens=None, speculative_draft_tensor_parallel_size=None, speculative_max_model_len=None, speculative_disable_by_batch_size=None, ngram_prompt_lookup_max=None, ngram_prompt_lookup_min=None, spec_decoding_acceptance_method='rejection_sampler', typical_acceptance_sampler_posterior_threshold=None, typical_acceptance_sampler_posterior_alpha=None, model_loader_extra_config=None, preemption_mode=None, served_model_name=['Qwen2-72B-Instruct'], qlora_adapter_name_or_path=None, otlp_traces_endpoint=None, engine_use_ray=False, disable_log_requests=False, max_log_len=None)
WARNING 07-19 06:57:49 config.py:1378] Casting torch.bfloat16 to torch.float16.
2024-07-19 06:57:49,153 INFO worker.py:1603 -- Connecting to existing Ray cluster at address: 23.248.165.186:6379...
2024-07-19 06:57:49,160 INFO worker.py:1788 -- Connected to Ray cluster.
INFO 07-19 06:57:50 llm_engine.py:174] Initializing an LLM engine (v0.5.2) with config: model='/model/models/Qwen2-72B-Instruct', speculative_config=None, tokenizer='/model/models/Qwen2-72B-Instruct', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, rope_scaling=None, rope_theta=None, tokenizer_revision=None, trust_remote_code=True, dtype=torch.float16, max_seq_len=32768, download_dir=None, load_format=LoadFormat.AUTO, tensor_parallel_size=16, pipeline_parallel_size=1, disable_custom_all_reduce=True, quantization=None, enforce_eager=True, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='outlines'), observability_config=ObservabilityConfig(otlp_traces_endpoint=None), seed=0, served_model_name=Qwen2-72B-Instruct, use_v2_block_manager=False, enable_prefix_caching=False)
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
INFO 07-19 06:58:24 selector.py:150] Cannot use FlashAttention-2 backend for Volta and Turing GPUs.
INFO 07-19 06:58:24 selector.py:53] Using XFormers backend.
(RayWorkerWrapper pid=13073, ip=23.248.165.187) INFO 07-19 06:58:24 selector.py:150] Cannot use FlashAttention-2 backend for Volta and Turing GPUs.
(RayWorkerWrapper pid=13073, ip=23.248.165.187) INFO 07-19 06:58:24 selector.py:53] Using XFormers backend.
INFO 07-19 06:58:26 utils.py:737] Found nccl from library libnccl.so.2
INFO 07-19 06:58:26 pynccl.py:63] vLLM is using nccl==2.20.5
(RayWorkerWrapper pid=12911, ip=23.248.165.187) INFO 07-19 06:58:26 utils.py:737] Found nccl from library libnccl.so.2
(RayWorkerWrapper pid=12911, ip=23.248.165.187) INFO 07-19 06:58:26 pynccl.py:63] vLLM is using nccl==2.20.5
(RayWorkerWrapper pid=12911, ip=23.248.165.187) Exception ignored in: <function ShmRingBuffer.__del__ at 0x7f41b057bc40>
(RayWorkerWrapper pid=12911, ip=23.248.165.187) Traceback (most recent call last):
(RayWorkerWrapper pid=12911, ip=23.248.165.187)   File "/model/anaconda3/envs/vllm/lib/python3.11/site-packages/vllm/distributed/device_communicators/shm_broadcast.py", line 122, in __del__
(RayWorkerWrapper pid=12911, ip=23.248.165.187)     self.shared_memory.close()
(RayWorkerWrapper pid=12911, ip=23.248.165.187)     ^^^^^^^^^^^^^^^^^^
(RayWorkerWrapper pid=12911, ip=23.248.165.187) AttributeError: 'ShmRingBuffer' object has no attribute 'shared_memory'
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340] Error executing method init_device. This might cause deadlock in distributed execution.
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340] Traceback (most recent call last):
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]   File "/model/anaconda3/envs/vllm/lib/python3.11/site-packages/vllm/worker/worker_base.py", line 332, in execute_method
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]     return executor(*args, **kwargs)
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]            ^^^^^^^^^^^^^^^^^^^^^^^^^
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]   File "/model/anaconda3/envs/vllm/lib/python3.11/site-packages/vllm/worker/worker.py", line 132, in init_device
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]     init_worker_distributed_environment(self.parallel_config, self.rank,
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]   File "/model/anaconda3/envs/vllm/lib/python3.11/site-packages/vllm/worker/worker.py", line 346, in init_worker_distributed_environment
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]     ensure_model_parallel_initialized(parallel_config.tensor_parallel_size,
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]   File "/model/anaconda3/envs/vllm/lib/python3.11/site-packages/vllm/distributed/parallel_state.py", line 923, in ensure_model_parallel_initialized
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]     initialize_model_parallel(tensor_model_parallel_size,
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]   File "/model/anaconda3/envs/vllm/lib/python3.11/site-packages/vllm/distributed/parallel_state.py", line 889, in initialize_model_parallel
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]     _TP = init_model_parallel_group(group_ranks,
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]   File "/model/anaconda3/envs/vllm/lib/python3.11/site-packages/vllm/distributed/parallel_state.py", line 732, in init_model_parallel_group
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]     return GroupCoordinator(
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]            ^^^^^^^^^^^^^^^^^
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]   File "/model/anaconda3/envs/vllm/lib/python3.11/site-packages/vllm/distributed/parallel_state.py", line 197, in __init__
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]     self.mq_broadcaster = MessageQueue.create_from_process_group(
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]   File "/model/anaconda3/envs/vllm/lib/python3.11/site-packages/vllm/distributed/device_communicators/shm_broadcast.py", line 484, in create_from_process_group
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]     dist.broadcast_object_list(recv,
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]   File "/model/anaconda3/envs/vllm/lib/python3.11/site-packages/torch/distributed/c10d_logger.py", line 75, in wrapper
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]     return func(*args, **kwargs)
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]            ^^^^^^^^^^^^^^^^^^^^^
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]   File "/model/anaconda3/envs/vllm/lib/python3.11/site-packages/torch/distributed/distributed_c10d.py", line 2674, in broadcast_object_list
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]     object_list[i] = _tensor_to_object(obj_view, obj_size, group)
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]   File "/model/anaconda3/envs/vllm/lib/python3.11/site-packages/torch/distributed/distributed_c10d.py", line 2362, in _tensor_to_object
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]     return _unpickler(io.BytesIO(buf)).load()
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]   File "/model/anaconda3/envs/vllm/lib/python3.11/site-packages/vllm/distributed/device_communicators/shm_broadcast.py", line 111, in __init__
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]     self.shared_memory = shared_memory.SharedMemory(name=name)
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]   File "/model/anaconda3/envs/vllm/lib/python3.11/multiprocessing/shared_memory.py", line 104, in __init__
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]     self._fd = _posixshmem.shm_open(
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340]                ^^^^^^^^^^^^^^^^^^^^^
(RayWorkerWrapper pid=12911, ip=23.248.165.187) ERROR 07-19 06:58:27 worker_base.py:340] FileNotFoundError: [Errno 2] No such file or directory: '/psm_871ece2c'

### PR DESCRIPTION
**Pull request by youkaichao**
fixes https://github.com/vllm-project/vllm/issues/6564

### CODE REVIEW

File: vllm/distributed/device_communicators/shm_broadcast.py [hunk 1]
```diff
@@ -108,8 +108,14 @@ def __init__(self,
             # created by the process. The following patch is a workaround.
             with patch("multiprocessing.resource_tracker.register",
                        lambda *args, **kwargs: None):
-                self.shared_memory = shared_memory.SharedMemory(name=name)
-            assert self.shared_memory.size == self.total_bytes_of_buffer
+                try:
+                    self.shared_memory = shared_memory.SharedMemory(name=name)
+                    assert self.shared_memory.size == self.total_bytes_of_buffer  # noqa
```
Comment (WoosukKwon) [inline vllm/distributed/device_communicators/shm_broadcast.py:113]: qq: what is this `noqa` for?

**Review by KuntaiDu — APPROVED** [review]

**Comment by njhill** [comment]
@youkaichao wouldn't the right thing to be to not attempt to use `ShmRingBuffer` at all for TP groups that span nodes?

File: vllm/distributed/device_communicators/shm_broadcast.py (hunk 1 above)
Comment (youkaichao) [reply vllm/distributed/device_communicators/shm_broadcast.py:113]: longer than 79 chars

**Comment by youkaichao** [comment]
@njhill even for a group that spans nodes, we use shm for the ranks inside the node, and zmq for ranks outside the node.

this is, actually, in preparation for the future, the alternative to ray dag implementation.

File: vllm/distributed/device_communicators/shm_broadcast.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/distributed/device_communicators/shm_broadcast.py:113]: This is super minor, but then you can parentheses
```suggestion
                    assert (self.shared_memory.size == self.total_bytes_of_buffer)
```

Comment (youkaichao) [reply vllm/distributed/device_communicators/shm_broadcast.py:113]: let's do it later, too busy rn :(

**Pull request merged**